### DISCUSSION TITLE
Custom objective function not calculated in parallel

### DISCUSSION THREAD

**Discussion by albisb** [mdolab/dafoam#504] [Bug] [answered]
Hi all,
I am testing a custom objective function with which I am not getting the results I was expecting. Therefore I started to debug adding printouts to the objective function and I realized that there is a forAll loop that is not being entered. More specifically:
forAll(objFuncFaceSources,idx)
        {
                const label& objFuncFaceI = objFuncFaceSources[idx];
                label bFaceI = objFuncFaceI - daIndex_.nLocalInternalFaces;
                const label patchI = daIndex_.bFacePatchI[bFaceI];
                const label faceI = daIndex_.bFaceFaceI[bFaceI];

                vector cellCenter = Sfb[patchI][faceI];
                vector dist = position - cellCenter;

                scalar weight = rbf(dist,scaleRBF_,sigma_);
                Info << "position " << dist << endl;
                Info << "pressure " << p.boundaryField()[patchI][faceI] << endl;
                interpolatedValue += p.boundaryField()[patchI][faceI] * weight;
                totalWeight += weight;
        }

Here, the way to call p and Sfb are done the same way as the center of pressure objective function. Since I didn't understand why this happens, I added printouts to the DAObjFunctionForce.C:

    Info << "objFuncValue: " << objFuncValue << endl;
    // calculate discrete force for each objFuncFace
    forAll(objFuncFaceSources, idxI)
    {
        const label& objFuncFaceI = objFuncFaceSources[idxI];
        label bFaceI = objFuncFaceI - daIndex_.nLocalInternalFaces;
        const label patchI = daIndex_.bFacePatchI[bFaceI];
        const label faceI = daIndex_.bFaceFaceI[bFaceI];

        Info << "Pressure in idxI " << idxI << ": " << p.boundaryField()[patchI][faceI] << endl;

        // normal force
        vector fN(Sfb[patchI][faceI] * p.boundaryField()[patchI][faceI]);
        // tangential force
        vector fT(Sfb[patchI][faceI] & devRhoReffb[patchI][faceI]);
        // project the force to forceDir
        objFuncFaceValues[idxI] = scale_ * ((fN + fT) & forceDir_);

        Info << "fN " << fN << endl;
        Info << "Face value: " << objFuncFaceValues[idxI] << endl;
        Info << "Coeff: " << objFuncValue << endl;

        objFuncValue += objFuncFaceValues[idxI];
    }

    Info << "objFuncValue: " << objFuncValue << endl;
And here I realize that it is also not entering the forAll loop presumably. This is what I get from the log when using runPrimal:
Time = 226
Minimal residual 9.979074643463256e-05 satisfied the prescribed tolerance 0.0001

objFuncFaceSources: 0()
objFuncValue: 0
objFuncValue: 0
CL-part1-force: 0.1573367480219689
Printing Primal Residual Statistics.
U Residual Norm2: (0.03327061495743294 0.0120346896526651 1.918414147281758e-13)
U Residual Mean: (0.000134383083956075 4.119950178144015e-05 6.237689130617107e-16)
U Residual Max: (0.002591793141887858 0.001222042055753093 4.396322487286621e-14)
p Residual Norm2: 0.00250613692416529
p Residual Mean: 8.661147860752775e-06
p Residual Max: 0.0003681159374928274
nuTilda Residual Norm2: 1.459538769317628e-06
nuTilda Residual Mean: 1.466226671326751e-09
nuTilda Residual Max: 2.018154636569081e-07
phi Residual Norm2: 0.0002468435543236913
phi Residual Mean: 2.940955414926608e-07
phi Residual Max: 2.065641472978707e-05
End

objFuncFaceSources: 0()
objFuncValue: 0
objFuncValue: 0
objFuncFaceSources: 0()
objFuncValue: 0
objFuncValue: 0
Finalising parallel run

As you can see, the value of objFuncValue, as printed in the objective function is 0, and the printouts within the forAll loop are not printed, therefore, it very much seems like it is not entering the forAll loop. Still, a value of CL is printed in the log, which I also don't understand why and could be a whole different discussion.

Finally, I decided to test the same but this time running in serial, and I see that in serial the printouts are working, and same for my custom objective function. An example of the log in serial:
...
Pressure in idxI 49: -5.719571698990153
fN (0.0005988305209458396 0.01714826247105601 -0)
Face value: 0.00343039485603128
Coeff: 0.1468663152666436
Pressure in idxI 50: -6.056556505703015
fN (0.0006341123213979187 0.0181586010453731 -0)
Face value: 0.003631381054898216
Coeff: 0.1502967101226748
Pressure in idxI 51: -6.689483691480471
fN (0.0007003788420967231 0.02005622591627279 1.133242617360856e-20)
Face value: 0.004010534863071528
Coeff: 0.1539280911775731
Pressure in idxI 52: 22.05683291996153
fN (0.04133136842494169 -0.001443323189544509 -1.868286419529069e-20)
Face value: -0.0002776387396965273
Coeff: 0.1579386260406446
Pressure in idxI 53: 24.84447620904951
fN (0.04655501554765938 -0.001625736966623491 3.417560752412481e-17)
Face value: -0.0003242914743927957
Coeff: 0.1576609873009481
objFuncValue: 0.1573366958265553

Why is this happening? I can run in serial but it would be great to be able to run in parallel.

Thanks

**Comment by friedenhe** [comment, marked as answer]
The behavior you saw is expected because the `Info` command will print things on proc0 only. If proc0 does not own any objFunc patches, it will return 0. Please use `Pout` and ask every processor to print things.

**Reply by albisb** [reply]
You're very much right, it is a problem of the interpolation I am using that expects boundary data to be available, which is not the case for all the processors.
Thanks
